### PR TITLE
[Serve] [Dashboard] Initial PR for exporting Serve data to cluster snapshot

### PR DESCRIPTION
**Pull request by architkulkarni**
## Why are these changes needed?
Adds some Serve Deployments information to the cluster snapshot in the dashboard, alongside the existing Job and Actor information. 

TODOs for followup PR:
- Support non-detached Serve controllers.   Currently this only works with serve.start(detached=True)
- Deployment start time
- Deployment end time (Do we need this?  When deployments are deleted, they are removed from the deployments list.). On a related note, status is always "RUNNING".
- Some way of associating actors with replica tags.  EIther (1) add this metadata in the actor table, or (2) add a dict in the serve snapshot.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/controller.py [hunk 1]
```diff
@@ -119,9 +121,30 @@ async def run_control_loop(self) -> None:
                     self.backend_state.update()
                 except Exception as e:
                     logger.error(f"Exception updating backend state: {e}")
-
+            self._put_serve_snapshot()
```
Comment (architkulkarni) [inline python/ray/serve/controller.py:124]: This will be called 10x a second, is this reasonable?

File: python/ray/serve/api.py [hunk 2]
```diff
@@ -668,7 +668,7 @@ def start(
     except RayServeException:
         pass
 
-    # Try to get serve controller if it exists
+    controller_name = SERVE_CONTROLLER_NAME
```
Comment (edoakes) [inline python/ray/serve/api.py:671]: is this change intentional

File: python/ray/serve/controller.py (hunk 1 above)
Comment (edoakes) [reply python/ray/serve/controller.py:124]: Should be fine. Really we should reduce the update period. The reason it's tuned so low right now is because we have a bunch of different states that we need to checkpoint between so deploying something takes a few update periods. cc @jiaodong we can increase the update period after your refactoring here.

File: dashboard/modules/snapshot/snapshot_head.py [hunk 3]
```diff
@@ -80,10 +86,25 @@ async def get_actor_info(self):
                 "current_raylet_id": actor_table_entry.address.raylet_id.hex(),
                 "ip_address": actor_table_entry.address.ip_address,
                 "port": actor_table_entry.address.port,
+                "ray_replica_id": "TODO",
+                "metadata": {
+                    "deployment_id": "TODO"
+                }
```
Comment (edoakes) [inline dashboard/modules/snapshot/snapshot_head.py:92]: is it possible that we leave the data out here and instead just include it in the Serve endpoint (e.g., a dictionary of actorIDs to serve replicaIDs or something? that would be preferrable

**Review by edoakes — COMMENTED** [review]
Looks pretty reasonable to me pending the TODOs being filled in. @kombuchafox please also take a quick look at the data being exported.

File: python/ray/serve/api.py (hunk 2 above)
Comment (architkulkarni) [reply python/ray/serve/api.py:671]: Ah sorry, good catch

File: python/ray/experimental/internal_kv.py [hunk 4]
```diff
@@ -23,9 +23,11 @@ def _internal_kv_initialized():
 
 
 @client_mode_hook
-def _internal_kv_get(key: Union[str, bytes]) -> bytes:
+def _internal_kv_get(key: Union[str, bytes], gcs_client=None) -> bytes:
     """Fetch the value of a binary key."""
```
Comment (simon-mo) [inline python/ray/experimental/internal_kv.py:27]: update the comment?

File: python/ray/serve/controller.py [hunk 5]
```diff
@@ -119,9 +121,30 @@ async def run_control_loop(self) -> None:
...
             await asyncio.sleep(CONTROL_LOOP_PERIOD_S)
 
+    def _put_serve_snapshot(self) -> None:
+        val = []
+        for name, (backend_info,
+                   route_prefix) in self.list_deployments().items():
+            deployment_info = {}
+            deployment_info["name"] = name
+            deployment_info["namespace"] = "TODO"
+            deployment_info["ray_job_id"] = "TODO"
```
Comment (architkulkarni) [inline python/ray/serve/controller.py:134]: After discussing with Simon, with a naive implementation these fields will all be the same across all deployments; they'll just be the job_id and namespace of the Controller.  Is this okay?  Or do we want to somehow allow different job_id's and namespaces per Deployment (for example, if `.deploy()` is called from a different driver script, does one expect the resulting deployment to inherit the job id from that driver script?)

File: dashboard/modules/snapshot/snapshot_head.py (hunk 3 above)
Comment (kombuchafox) [reply dashboard/modules/snapshot/snapshot_head.py:92]: Having this here would be easier for ingestion logic imo.
But I see the benefit of having in the serve endpoint for isolation of business logic. 

The main disadvantage for ingestion is that I have to reorganize the actors data and do the association myself. This may not be that hard but it will make the code more complex to understand and maintain on the product side.

Comment (architkulkarni) [reply dashboard/modules/snapshot/snapshot_head.py:92]: Based on what we decided in our discussion in standup, I'll remove this. In a future PR I'll publish this information in the Serve cluster snapshot.

File: dashboard/modules/snapshot/tests/test_snapshot.py [hunk 6]
```diff
@@ -68,3 +68,43 @@ def ping(self):
...
+    pprint.pprint(data)
+    jsonschema.validate(instance=data, schema=json.load(open(schema_path)))
+
+    assert len(data["data"]["snapshot"]["deployments"]) == 1
+    for deployment in data["data"]["snapshot"]["deployments"]:
+        assert deployment["name"] == "my_func"
+        assert deployment["version"] == "v1"
+        assert deployment["httpRoute"] == "/my_func"
+        assert deployment["className"] == "my_func"
+        assert deployment["status"] == "RUNNING"
```
Comment (kombuchafox) [inline dashboard/modules/snapshot/tests/test_snapshot.py:109]: I take it this will not be static when end_time exists ?

**Review by kombuchafox — COMMENTED** [review]
Overall LGTM, thanks

File: dashboard/modules/snapshot/tests/test_snapshot.py (hunk 6 above)
Comment (architkulkarni) [reply dashboard/modules/snapshot/tests/test_snapshot.py:109]: That's right, once we persist the history of dead deployments, the status will be updated as well as the end time

**Comment by architkulkarni** [comment]
Travis build failing on master, //python/ray/tests:test_basic_3_client_mode  failing on master

**Review by edoakes — APPROVED** [review]

**Comment by edoakes** [comment]
@alanwguo @ijrsvt @joeybai @wuisawesome could one of you gracious code owners please rubber stamp this so I can click merge :)

File: python/ray/serve/controller.py [hunk 7]
```diff
@@ -119,9 +121,32 @@ async def run_control_loop(self) -> None:
...
+            deployment_info["version"] = backend_info.version or "Unversioned"
+            # TODO(architkulkarni): When we add the feature to allow
+            # deployments with no HTTP route, update the below line.
+            # Or refactor the route_prefix logic in the Deployment class now.
+            deployment_info["http_route"] = route_prefix or f"/{name}"
+            deployment_info["status"] = "RUNNING"
+            deployment_info["start_time"] = "TODO"
+            deployment_info["end_time"] = ""
+            val.append(deployment_info)
+        self.kv_store.put(SNAPSHOT_KEY, json.dumps(val))
```
Comment (kombuchafox) [inline python/ray/serve/controller.py:148]: Could we actually change this to be like the other fields: `actors` and `jobs` in which 
its `{"identifier": value}`
In this case it would be `{"deployment_name": value}`.

Comment (kombuchafox) [reply python/ray/serve/controller.py:148]: also @architkulkarni  start_time / end_time are planned to be unix timestamps in the future correct?

I think it will be better if we use `0` for start_time and end_time right now instead of a string so we don't have to change the type in the end.

The jobs and actors use `0` for when end_time is not known

Comment (architkulkarni) [reply python/ray/serve/controller.py:148]: This all sounds good, I had the same thought regarding the first point.  I'll include it in the followup PR if that's okay.  As for start/end times, it hasn't been implemented yet so we have freedom to choose there--your suggestion sounds good to me.

**Review by wuisawesome — COMMENTED** [review]
Can we add these new outputs to `snapshot_schema.json`?

**Comment by architkulkarni** [comment]
> Can we add these new outputs to `snapshot_schema.json`?

Ah good catch, I overlooked that.  I'll update it.  

I wonder why
```python
    schema_path = os.path.join(
        os.path.dirname(dashboard.__file__),
        "modules/snapshot/snapshot_schema.json")
    pprint.pprint(data)
    jsonschema.validate(instance=data, schema=json.load(open(schema_path)))
```
was passing without the update...

**Comment by wuisawesome** [comment]
I think adding extra values is OK by default since it doesn't break backwards compatibility. I don't have a strong opinion if we want to change that.

**Review by alanwguo — APPROVED** [review]

**Review by wuisawesome — APPROVED** [review]

**Comment by alanwguo** [comment]
I wasn't aware I'm a codeowners of dashboard code already. Haven't had the time to do any sort of knowledge transfer yet 

> I think adding extra values is OK by default since it doesn't break backwards compatibility. I don't have a strong opinion if we want to change that.

No need to change this. I think its the right behavior. However, if you don't update the schema, eventually, the product will not be able to use the data. That's because, we're going to generate typesafe classes based off the schema. This can be the forcing function that will keep the schema up to date

**Comment by architkulkarni** [comment]
Failing tests:

//python/ray/tests:test_basic_3_client_mode, //python/ray/tests:test_plasma_unlimited, failing on master
 //rllib:tests/test_gpus, flaky on master

//python/ray/tests:test_kill_raylet_signal_log on Windows is flaky on master
